[consensus] Implement Raft proxying

mpercy opened the pull request:
This stack implements Raft proxying, split into multiple parts:

* Logic for calculating routes
* Support for actually proxying received messages meant for another node
* Support for tracking proxied messages and sending messages through proxies at the leader level

There are a couple layers of tests included in these patches:

* Tests for the routing logic
* Tests for the proxying logic
* An end-to-end test to ensure that a topology with proxies defined will eventually distribute writes and converge.

mpercy committed:
[net_util] Disable UNIQUE_LOOPBACK in Facebook env

I'm running on a machine with 32-bit pids. This patch disables
UNIQUE_LOOPBACK -- a good patch would detect the use of 32-bit pids and
dynamically disable UNIQUE_LOOPBACK.

mpercy committed:
[consensus][proxy][3/n] add support for proxying consensus messages

This patch adds support for proxying consensus messages.

Currently supported:

* RPC to change proxy topology.
* PROXY_OP being translated to a real op if available and forwarded.
* Single-hop proxying support.
* Inefficient server-side proxying implementation (blocking).
* Specification of proxying topology in the Raft config definition.
* Multi-hop routing.
* Proxy timed-waiting if proxied op is not yet available.

Tests:

1. Simple test that acts like a leader and sends RPCs to a proxy an
   validates that a downstream node has received the RPC.
2. Same as the above, but also tests multi-hop routing.

TODO:

* Leader knowledge of proxies and the topology.
* TTL support to avoid routing loops.
* Non-blocking request processing.

Change-Id: I1b42aa4777df1858409d6933ae8316e0e36db4ae

mpercy committed:
[consensus][proxy][1/n] implement proxy routing logic

This patch implements a RoutingTable class to control the routes of
messages being proxied across a replica set topology. This routing table
is actually more of a routing tree, since there is a single-parent
relationship among nodes represented in the routing table, and so there
is only one possible route to each node at any given instance of the
configuration. That said, most people have some context for what a
routing table is used for, so that's why I chose that name. See the
comment at the top of routing.h for details on the assumptions made and
the behavior of the RoutingTable class.

This patch also implements a wrapper class called DurableRoutingTable
that persists the relevant routing metadata to disk.

The proxy routing topology is specified in a new protobuf called
ProxyTopologyPB, which is separate from RaftConfigPB so that the
topology can be changed independently of Raft config change. Such an RPC
API appears later in this patch series.

Included is a simple unit test of the next-hop routing behavior for
RoutingTable. The DurableRoutingTable could use unit tests, but for now
has integration test coverage in a later patch in this series.

Some TODOs:

* Could use some lower level test coverage for DRT including a
  persistence test and coverage for Incomplete status handling.

Change-Id: I44e9ccfe4ca97ac24c30b7163778840e1d4b823e

calid reviewed (approved):
lgtm

anirbanr-fb commented on src/kudu/consensus/routing.cc:
```diff
@@ -0,0 +1,156 @@
+namespace kudu {
+namespace consensus {
+
+Status RoutingTable::Init(RaftConfigPB& config,
```
shouldn't this be const reference. are we using move semantics?

mpercy replied:
Fixed

anirbanr-fb commented on src/kudu/consensus/metadata.proto:
```diff
@@ -104,6 +104,10 @@ message RaftPeerPB {
 
+  // Optional node to receive replicate messges from, instead of directly from
+  // the leader.
+  optional bytes proxy_from = 6;
```
Are we allowing RaftPeerPB to diverge on all nodes of the ring?
is automation/going to hit the nodes independently to set this

mpercy replied:
The proxy node definitions will be eventually consistent over the course of a config change operation.

anirbanr-fb replied:
An easy way to not have to go through consensus is to pull this out from RaftPeerPB and make a separate
protobu called PeerUUID->ProxyFrom map. And assume that its someone else's concern to thread that
map to you. should be relatively simple change.

anirbanr-fb commented on src/kudu/consensus/routing.cc:
```diff
@@ -0,0 +1,156 @@
+  return Status::OK();
+}
+
+Status RoutingTable::MergeTrees(
```
this function is missing comments.

mpercy replied:
Yes, it needs them. Fixed. Don't forget to look at the header file comments though.

mpercy replied:
Added some comments.

anirbanr-fb commented on src/kudu/consensus/routing.cc:
```diff
@@ -0,0 +1,156 @@
+  ConstructNextHopIndicesRec(trees.begin()->second.get());
+
+  index_ = std::move(index);
+  topology_root_ = std::move(trees.begin()->second);
```
is there going to be a persistent representation of the topology_root_?

mpercy replied:
No, the topology root is just either the leader or the ultimate proxy parent of the leader. Since it's relative to the leader, it's a transient thing.

mpercy committed:
[consensus][proxy][4/n] convert PeerProxy to shared ownership

This will allow us to implement a PeerProxyPool while sharing instances
with Peers, which will make it possible to add dynamic proxy support to
the Peers. Needed for leader proxying support.

mpercy commented on src/kudu/consensus/consensus_peers.cc:
```diff
@@ -281,10 +287,22 @@ void Peer::SendNextRequest(bool even_if_queue_empty) {
+
+  // TODO(mpercy): Instead of going through proxy_, use an actual proxy if needed.
+  // Or, simply calculate the proxy_ as the proxy_ we're supposed to go through
+  // when it's instantiated. So what happens when we reconfigure Kudu?
```
Oops, this comment should be deleted now.

anirbanr-fb commented on src/kudu/consensus/consensus_queue.cc:
```diff
@@ -688,6 +690,13 @@ Status PeerMessageQueue::RequestForPeer(const string& uuid,
   *needs_tablet_copy = false;
 
+  // If the next hop != the destination, we are sending these messages via a proxy.
+  string next_hop_uuid = routing_table_.NextHop(local_peer_pb_.permanent_uuid(), uuid);
```
seems like lock is not held here.  Do we need to use GetNextRoutingHopFromLeader?

mpercy replied:
Good catch! We already took and released the lock in this function, so I'll just move the call up there.

mpercy replied:
Fixed

anirbanr-fb commented on src/kudu/consensus/consensus_peers.cc:
```diff
@@ -128,6 +132,7 @@ Peer::Peer(RaftPeerPB peer_pb,
            PeerMessageQueue* queue,
+           PeerProxyPool* peer_proxy_pool,
            ThreadPoolToken* raft_pool_token,
            shared_ptr<PeerProxy> proxy,
```
there is overlap of terminology between Proxy in PeerProxy and proxy in PeerProxyPool.
They actually mean 2 different things. (One is a local vs remote RPC abstraction, i.e. if you call functions
in PeerProxy they become RPC calls). The other means to send messages through a different "instance"

mpercy replied:
This is correct. It's unfortunate but I don't think it's a good idea to go in and get rid of the existing usage of the Proxy term since it would be a huge diff. Do you have any suggestions?

anirbanr-fb replied:
Not suggesting change PeerProxy, but new code could use different terminology.

anirbanr-fb commented on src/kudu/consensus/routing.cc:
```diff
@@ -0,0 +1,162 @@
+    std::unordered_map<std::string, std::unique_ptr<Node>>* trees) {
+
+  // Construct the peers.
+  for (const RaftPeerPB& peer : config.peers()) {
```
the routing table is being built in memory it seems, and what it primarily depends on is proxy_from field.
It should be realtively easy to change peer.proxy_from()  to something like proxy_map_config.proxy_from(peer),
and with the same outcome as this code assumes.

mpercy replied:
This is done, along with a bunch of other changes we discussed offline that are detailed in the individual commit messages in this stack.

mpercy committed:
[consensus][proxy][2/n] add DurableRoutingTable to CMeta Manager

This simplifies plumbing the DRT into RaftConsensus and provides a
pattern for integrating a durable leader ban as well, later on.

mpercy committed:
[consensus][proxy][5/n] leader support

This patch adds leader support for proxying, meaning that when a node is
a leader it will be proxy-aware and route messages according to proxying
rules.

Includes an integration test.
